### DISCUSSION TITLE
Kong for Kubernetes config generation improvements

### DISCUSSION THREAD

**Discussion by develohpanda** [Kong/insomnia#3752] [Ideas]
- [x] If multiple `Ingress` documents exist, they must have unique names. Otherwise, running `kubectl` will overwrite the previous one - #2163
- [ ] Consider adding [`konghq.com/protocols`](https://github.com/Kong/kubernetes-ingress-controller/blob/master/docs/references/annotations.md#konghqcomprotocols) annotation explicitly, to retain more routing information if a service can be both `http` and `https`.
  - Currently, a server being http or https doesn't do anything to the ingress document (as the protocol is stripped). It is only the existence of a `tls` section that modifies the Ingress document. https://github.com/Kong/insomnia/blob/75a129c0388a7fc639a423ff9460de35c239dd30/packages/openapi-2-kong/src/kubernetes/index.js#L160-L175
  - How should the annotation be controlled? What decides whether the annotation says `http`, or `http,https`, or `https`?
- [ ] Generate a k8s service manifest with unique names
  - a service can be for [multiple ports](https://kubernetes.io/docs/concepts/services-networking/service/#multi-port-services) (ie. support both `http` and `https`)
  - The `konghq.com/plugins` annotation can be applied to a service specifically. If the same plugin is applied to a service and the ingress document, which takes precendence? (If this behaves as expected, our conversion implementation can be simplified).
- [ ] Abstract logic from an Ingress resource to a k8s service manifest
  - create two k8s services, one for production and one for developer (each service in `OAS.services`)
  - add plugin annotation on the appropriate service
  - create a single ingress resource for each path, with two rules - one for each service
  - We can also add [`konghq.com/path`](https://github.com/Kong/kubernetes-ingress-controller/blob/master/docs/references/annotations.md#konghqcompath) on a service, as a routing prefix 🤔 
    - eg `konghq.com/path = "/ds-api"`
- [ ] Improve naming of KongPlugin resource - eg `{service}-{plugin-name}-{suffix}`
- [ ] Remove closing wildcard (unless it is required as part of a service variable)

**Comment by nijikokun** [comment]
> * How should the annotation be controlled? What decides whether the annotation says `http`, or `http,https`, or `https`?

Server URL protocol. We should use `http` or `https`, not both.

> * Currently, a server being http or https doesn't do anything to the ingress document (as the protocol is stripped). It is only the existence of a `tls` section that modifies the Ingress document.

True. This modifies the K8s behavior, which would be nice if K4K8s converted this to annotations for Kong. The `http` / `https` annotation would only impact Kong not K8s.

> * We can also add [`konghq.com/path`](https://github.com/Kong/kubernetes-ingress-controller/blob/master/docs/references/annotations.md#konghqcompath) on a service, as a routing prefix

My understanding was that K4K8s did this under the hood, if that's not the case, this would be ideal.

> * The `konghq.com/plugins` annotation can be applied to a service specifically. If the same plugin is applied to a service and the ingress document, which takes precendence? (If this behaves as expected, our conversion implementation can be simplified).

Let's meet tomorrow to discuss this in depth

**Comment by develohpanda** [comment]
> Server URL protocol. We should use `http` or `https`, not both.

What about a service that is defined as:
```yaml
services:
  - url: {scheme}://api.insomnia.rest
    variables:
      scheme:
        default: 'http'
        enum:
          - 'http'
          - 'https'
```
